### PR TITLE
delete IR for non-inlineable functions after codegen to save memory

### PR DESCRIPTION
**Pull request by JeffBezanson**
This saves an appreciable amount of memory.

Before:

```
    From worker 3:       * linalg/qr             in 109.82 seconds, maxrss  381.65 MB
    From worker 3:       * linalg/dense          in  41.32 seconds, maxrss  439.55 MB
    From worker 3:       * linalg/matmul         in 136.85 seconds, maxrss  516.54 MB
    From worker 3:       * linalg/schur          in   5.81 seconds, maxrss  521.65 MB
    From worker 3:       * linalg/special        in   3.30 seconds, maxrss  524.22 MB
    From worker 3:       * linalg/eigen          in  28.30 seconds, maxrss  537.02 MB
    From worker 3:       * linalg/bunchkaufman   in  22.59 seconds, maxrss  542.05 MB
    From worker 3:       * linalg/svd            in  12.16 seconds, maxrss  551.03 MB
    From worker 3:       * linalg/lapack         in  26.47 seconds, maxrss  556.81 MB
    From worker 2:       * linalg/triangular     in 398.41 seconds, maxrss  827.11 MB
```

with this change:

```
    From worker 3:       * linalg/qr             in  99.19 seconds, maxrss  380.03 MB
    From worker 3:       * linalg/dense          in  49.18 seconds, maxrss  421.14 MB
    From worker 3:       * linalg/matmul         in 150.76 seconds, maxrss  487.86 MB
    From worker 3:       * linalg/schur          in   6.09 seconds, maxrss  487.93 MB
    From worker 3:       * linalg/special        in   2.99 seconds, maxrss  487.93 MB
    From worker 3:       * linalg/eigen          in  27.72 seconds, maxrss  491.68 MB
    From worker 3:       * linalg/bunchkaufman   in  22.36 seconds, maxrss  494.21 MB
    From worker 3:       * linalg/svd            in  11.87 seconds, maxrss  501.07 MB
    From worker 3:       * linalg/lapack         in  28.59 seconds, maxrss  519.14 MB
    From worker 2:       * linalg/triangular     in 408.06 seconds, maxrss  775.71 MB
```

### CODE REVIEW

**Comment by vtjnash** [comment]
This seems to be barely a couple percent savings? That seems rather disappointing. Is the large slowdown (esp the 50% on linalg/dense) real or just the usual travis noise?

**Comment by JeffBezanson** [comment]
Ah, I needed to fix the inlining logic a bit, and now the numbers look better.

**Comment by JeffBezanson** [comment]
@vtjnash - I think this is ready for final review.

**Comment by vtjnash** [comment]
lgtm

**Comment by tkelman** [comment]
This will need to be reverted if you can't propose a fix for #16921.

**Comment by JeffBezanson** [comment]
Ok, ok, give me some time.

**Comment by JeffBezanson** [comment]
Also, if necessary, you can change `JL_DELETE_NON_INLINEABLE` to 0 in options.h instead of reverting.

File: src/codegen.cpp [hunk 1]
```diff
@@ -874,6 +874,16 @@ static void to_function(jl_lambda_info_t *li)
     // mark the pointer calling convention
     li->jlcall_api = (f->getFunctionType() == jl_func_sig ? 0 : 1);
 
+    // if not inlineable, code won't be needed again
+    if (JL_DELETE_NON_INLINEABLE &&
+        li->def && li->inferred && !li->inlineable && !jl_options.outputji) {
+        li->code = jl_nothing;
+        li->slottypes = jl_nothing;
```
Comment (Keno) [inline src/codegen.cpp]: Can we keep slottypes? The debugger uses them.

Comment (JeffBezanson) [inline src/codegen.cpp]: OK. You can just go ahead and take this line out.

**Pull request merged**